### DISCUSSION TITLE
Clarification about Builtins in docs/pages/libraries.md

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#83] [Q&A] [answered]
I was reading https://github.com/edubart/nelua-lang/blob/master/docs/pages/libraries.md to see whether `string.format()` is considered builtin or not and could not distinguish the clarification.

In Lua I can do `local name = "stefanos"; print(string.format("My name is %s", name))` and works without the need for calling `require`; in Nelua I'm forced to use `require` as it throws error messages.

![image](https://user-images.githubusercontent.com/1777466/126327725-2e0550aa-6ab9-42cb-8aa2-1935ef62992a.png)

Can we make clearer which functions are actually builtin and which are not?

A label next to each function will work just fine, like for example _`require (builtin)`_, _`string.format() (library)`_

**Comment by edubart** [comment, marked as answer]
The `string` is builtin type, but methods for it are not, because they generate runtime code. In Nelua you usually need to require a module for APIs that generates runtime code, because what runtime to include is the user choice by design.

> Can we make clearer which functions are actually builtin and which are not?

The documentation already state what functions are builtin, that is, just the functions from "builtins" section from that page. Any other function or method not there will be only available through a module imported with `require`. Though many modules like `io` already requires `'string'`. Aside from those builtin functions there are builtin types, like `int64`, `float64`, etc, and `string` is also one of the builtin types, just not its methods.

I know it's not ideal the current situation, it's a very common mistake of new Nelua users to attempt to use string methods while forgetting to do `require 'string'`, and they find that odd and don't like it, but I haven't a good solution for this yet.

**Comment by stefanos82** [comment]
I see.

Can we add a suggestion like GCC's gives us when we attempt to use a function that its header file has not being included?

![image](https://user-images.githubusercontent.com/1777466/126354548-af6ad665-e963-440e-9bb1-543e71cf3a4b.png)

**Reply by edubart** [reply]
Good idea! I will do this!

**Reply by edubart** [reply]
I've implemented this in latest commit, thanks for the suggestion. Now on attempt to use any global symbol that Nelua standard library provides, the user will get a suggestion on the needed module to `require` to fix the error.

**Reply by stefanos82** [reply]
Very nice! Let me check it, I will let you know.

**Reply by stefanos82** [reply]
Perfect! It works wonders :partying_face:

**Comment by stefanos82** [comment]
Question: how linting flag supposed to work?

In my case, should not make sense to throw a list of warnings for missing module(s)?

**Reply by edubart** [reply]
The `--lint` options just check the syntax of the called file, the more interesting for checking for errors is the `--analyze` option.
At this moment the Nelua compiler is designed in such a way that on the first error it will exit, for both options, so it's not possible to generate a list of errors. This may be changed in the future but right now I've no motivation for doing so.

**Reply by stefanos82** [reply]
Is `--analyze` used by default during execution of `nelua` command?

As far as I can see, the output for `nelua --analyze foo.nelua` and `nelua foo.nelua` is exactly the same, that's why I ask.

**Reply by edubart** [reply]
> Is --analyze used by default during execution of nelua command?

Yes, Nelua always has to analyze before compiling, the `--analyze` just makes the compiler stop after analyzing, so C code generation, binary compilation and execution is not performed with `--analyze`.

**Reply by stefanos82** [reply]
Ah OK.

About resolving missing modules, Golang has `gopls` which is a useful Language Server Protocol (LSP); it helps developers to reduce such issues even further.

I honestly like it when I type something like `fmt.Printf()` and as soon as I save changes, it adds the missing module on top of my source file.

Maybe such a lightweight mechanism could be added in Nelua without the need for LSP and be used in the form of a flag, something like `nelua --resolve-missing-modules`? :thinking:

**Reply by edubart** [reply]
> Maybe such a lightweight mechanism could be added in Nelua without the need for LSP and be used in the form of a flag, something like `nelua --resolve-missing-modules`?

The commit https://github.com/edubart/nelua-lang/commit/03709e2f2ef083cc41e688baf9d6975cfa63f440 does something like that, it will suggest modules to be required when using any method from the standard library.

**Reply by stefanos82** [reply]
Ah yes, I have already seen that in action after we discussed it back in July.
